mqtt/mqtt-bridge/src/client.rs

```diff
@@ -184,6 +186,36 @@ impl BridgeIoSource {
     }
 }
 
+#[derive(Debug, Clone)]
+pub struct InFlightPublishHandle {
+    publish_handle: PublishHandle,
+    permits: Arc<Semaphore>,
+}
+
+impl InFlightPublishHandle {
+    fn new(publish_handle: PublishHandle, permits: Arc<Semaphore>) -> Self {
+        Self {
+            publish_handle,
+            permits,
+        }
+    }
+
+    pub async fn publish(&self, publication: Publication) {
+        let mut publish_handle = self.publish_handle.clone();
+        let permits = self.permits.clone();
+
+        let fut = async move {
+            let permit = permits.acquire().await;
```

Comment:
I think this needs to be outside the async block. If it stays here then all calls to publish will spawn a new task waiting for a permit. Unfortunately, I think that this is hard because of the above-mentioned lifetime issue.